File: ImperatorToCK3/Source/Mappers/NicknameMapper/NicknameMapper.cpp
#include "NicknameMapper.h"
#include "NicknameMapping.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



mappers::NicknameMapper::NicknameMapper() {
	Log(LogLevel::Info) << "-> Parsing nickname mappings.";
	registerKeys();
	parseFile("configurables/nickname_map.txt");
	clearRegisteredKeywords();
	Log(LogLevel::Info) << "<> Loaded " << impToCK3NicknameMap.size() << " nickname links.";
}


mappers::NicknameMapper::NicknameMapper(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void mappers::NicknameMapper::registerKeys() {
	registerKeyword("link", [this](std::istream& theStream) {
		const NicknameMapping theMapping(theStream);
		for (const auto& imperatorNickname: theMapping.impNicknames) {
			if (theMapping.ck3Nickname)
				impToCK3NicknameMap.emplace(imperatorNickname, *theMapping.ck3Nickname);
		}
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


std::optional<std::string> mappers::NicknameMapper::getCK3NicknameForImperatorNickname(const std::string& impNickname) const
{
	const auto& mapping = impToCK3NicknameMap.find(impNickname);
	if (mapping != impToCK3NicknameMap.end())
		return mapping->second;
	return std::nullopt;
}

File: ImperatorToCK3/Mappers/Nickname/NicknameMapper.cs
using commonItems;
using System.Collections.Generic;

namespace ImperatorToCK3.Mappers.Nickname;

public sealed class NicknameMapper {
	private readonly Dictionary<string, string> impToCK3NicknameMap = new();

	public NicknameMapper() { }
	public NicknameMapper(string filePath) {
		Logger.Info("Parsing nickname mappings...");
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseFile(filePath);
		Logger.Info($"Loaded {impToCK3NicknameMap.Count} nickname links.");

		Logger.IncrementProgress();
	}
	public NicknameMapper(BufferedReader reader) {
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
	}
	private void RegisterKeys(Parser parser) {
		parser.RegisterKeyword("link", reader => {
			var mapping = new NicknameMapping(reader);
			if (mapping.CK3Nickname is null) {
				return;
			}

			foreach (var imperatorNickname in mapping.ImperatorNicknames) {
				impToCK3NicknameMap.Add(imperatorNickname, mapping.CK3Nickname);
			}
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
	}
	public string? GetCK3NicknameForImperatorNickname(string? impNickname) {
		if (impNickname is null) {
			return null;
		}
		return impToCK3NicknameMap.TryGetValue(impNickname, out var ck3Nickname) ? ck3Nickname : null;
	}
}